File: ios/Approach/Sources/StatisticsLibrary/Protocols/AveragingStatistic.swift
import Charts
import StringsLibrary

public protocol AveragingStatistic: Statistic {
	var total: Int { get set }
	var divisor: Int { get set }
	var average: Double { get }
}

extension AveragingStatistic {
	public static var supportsAggregation: Bool {
		false
	}

	public static var supportsWidgets: Bool {
		true
	}

	public var average: Double {
		divisor > 0 ? Double(total) / Double(divisor) : 0
	}

	public mutating func aggregate(with: Statistic) {
		guard let with = with as? Self else { return }
		self.total += with.total
		self.divisor += with.divisor
	}

	public var formattedValue: String {
		format(average: average)
	}

	public var formattedValueDescription: String? {
		nil
	}

	public var isEmpty: Bool {
		divisor == 0
	}
}

File: android/core/statistics/src/main/java/ca/josephroque/bowlingcompanion/core/statistics/interfaces/AveragingStatistic.kt
package ca.josephroque.bowlingcompanion.core.statistics.interfaces

import ca.josephroque.bowlingcompanion.core.statistics.Statistic
import ca.josephroque.bowlingcompanion.core.statistics.utils.formatAsAverage

interface AveragingStatistic: Statistic {
	var total: Int
	var divisor: Int

	override val supportsAggregation: Boolean
		get() = false

	override val supportsWidgets: Boolean
		get() = true

	val average: Double
		get() = if (divisor == 0) 0.0 else total.toDouble() / divisor.toDouble()

	override val formattedValue: String
		get() = average.formatAsAverage

	override val isEmpty: Boolean
		get() = divisor == 0

	override fun aggregateWithStatistic(statistic: Statistic) {
		if (statistic is AveragingStatistic) {
			total += statistic.total
			divisor += statistic.divisor
		}
	}
}